pkg/controller/telemetry/metric.go

```diff
@@ -429,36 +476,124 @@ func buildPrincipal(workload *workloadapi.Workload) string {
+		}
+		if data.success != connection_success {
+			m.metricCache.WorkloadConnFailed[labels] = m.metricCache.WorkloadConnFailed[labels] + 1
+		}
+		m.metricCache.WorkloadConnReceivedBytes[labels] = m.metricCache.WorkloadConnReceivedBytes[labels] + float64(data.receivedBytes)
+		m.metricCache.WorkloadConnSentBytes[labels] = m.metricCache.WorkloadConnSentBytes[labels] + float64(data.sentBytes)
+	} else {
+		if data.state == TCP_ESTABLISHED {
+			m.metricCache.WorkloadConnOpened[labels] = 1
+		}
+		if data.state == TCP_CLOSTED {
+			m.metricCache.WorkloadConnClosed[labels] = 1
+		}
+		if data.success != connection_success {
+			m.metricCache.WorkloadConnFailed[labels] = 1
+		}
+		m.metricCache.WorkloadConnReceivedBytes[labels] = float64(data.receivedBytes)
+		m.metricCache.WorkloadConnSentBytes[labels] = float64(data.sentBytes)
 	}
-	if data.success != connection_success {
-		tcpConnectionFailedInWorkload.With(commonLabels).Add(float64(1))
+}
+
+func (m *MetricController) buildServiceMetricsToPrometheus(data requestMetric, labels serviceMetricLabels) {
+	_, ok := m.metricCache.ServiceConnReceivedBytes[labels]
+	if ok {
+		if data.state == TCP_ESTABLISHED {
+			m.metricCache.ServiceConnOpened[labels] = m.metricCache.ServiceConnOpened[labels] + 1
+		}
+		if data.state == TCP_CLOSTED {
+			m.metricCache.ServiceConnClosed[labels] = m.metricCache.ServiceConnClosed[labels] + 1
+		}
+		if data.success != connection_success {
+			m.metricCache.ServiceConnFailed[labels] = m.metricCache.ServiceConnFailed[labels] + 1
+		}
+		m.metricCache.ServiceConnReceivedBytes[labels] = m.metricCache.ServiceConnReceivedBytes[labels] + float64(data.receivedBytes)
+		m.metricCache.ServiceConnSentBytes[labels] = m.metricCache.ServiceConnSentBytes[labels] + float64(data.sentBytes)
+	} else {
+		if data.state == TCP_ESTABLISHED {
+			m.metricCache.ServiceConnOpened[labels] = 1
+		}
+		if data.state == TCP_CLOSTED {
+			m.metricCache.ServiceConnClosed[labels] = 1
+		}
+		if data.success != connection_success {
+			m.metricCache.ServiceConnFailed[labels] = 1
+		}
+		m.metricCache.ServiceConnReceivedBytes[labels] = float64(data.receivedBytes)
+		m.metricCache.ServiceConnSentBytes[labels] = float64(data.sentBytes)
 	}
-	tcpReceivedBytesInWorkload.With(commonLabels).Add(float64(data.receivedBytes))
-	tcpSentBytesInWorkload.With(commonLabels).Add(float64(data.sentBytes))
 }
 
-func buildServiceMetricsToPrometheus(data requestMetric, labels serviceMetricLabels) {
-	commonLabels := struct2map(labels)
+func (m *MetricController) updatePrometheusMetric() error {
+	m.mutex.RLock()
+	metricInfo := m.metricCache
+	m.mutex.RUnlock()
 
-	if data.state == TCP_ESTABLISHED {
-		tcpConnectionOpenedInService.With(commonLabels).Add(float64(1))
-	}
-	if data.state == TCP_CLOSTED {
-		tcpConnectionClosedInService.With(commonLabels).Add(float64(1))
+	val := reflect.ValueOf(metricInfo)
+	typ := reflect.TypeOf(metricInfo)
```

Comment:
since below you are still enumerating all the fields names, It seems that using reflect has no much benefit